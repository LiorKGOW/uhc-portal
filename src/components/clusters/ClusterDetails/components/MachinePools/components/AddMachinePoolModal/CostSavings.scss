.ocm-c-cost-savings-price-radios {
  padding-left: var(--pf-global--spacer--lg);
  padding-top: var(--pf-global--spacer--md);
}

.ocm-c-cost-savings-max-price__number_input {
  padding-left: var(--pf-global--spacer--2xl);
  white-space: nowrap;
}
