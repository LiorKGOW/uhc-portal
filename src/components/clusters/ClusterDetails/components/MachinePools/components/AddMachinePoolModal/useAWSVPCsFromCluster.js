import { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { getAWSCloudProviderVPCs } from '~/components/clusters/CreateOSDPage/CreateOSDWizard/ccsInquiriesActions';

/**
 * React hook fetching VPCs for a node pool. Request args extracted from cluster in global redux state.
 * @returns current vpcs state.
 */
export const useAWSVPCsFromCluster = () => {
  const dispatch = useDispatch();
  const cluster = useSelector((state) => state.clusters.details.cluster);
  const vpcs = useSelector((state) => state.ccsInquiries.vpcs);

  useEffect(() => {
    const subnet =
      cluster.aws && cluster.aws.subnet_ids && cluster.aws.subnet_ids[0]
        ? cluster.aws.subnet_ids[0]
        : undefined;
    dispatch(
      getAWSCloudProviderVPCs(
        { sts: { role_arn: cluster.aws.sts.role_arn } },
        cluster.region.id,
        subnet,
      ),
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return vpcs;
};
