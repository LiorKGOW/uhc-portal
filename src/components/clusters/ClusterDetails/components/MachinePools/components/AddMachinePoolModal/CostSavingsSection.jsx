import React from 'react';
import { FormGroup, GridItem, NumberInput, Alert } from '@patternfly/react-core';
import { Field } from 'redux-form';

import './CostSavings.scss';
import PropTypes from 'prop-types';
import ReduxCheckbox from '../../../../../../common/ReduxFormComponents/ReduxCheckbox';
import RadioButtons from '../../../../../../common/ReduxFormComponents/RadioButtons';
import ExternalLink from '../../../../../../common/ExternalLink';
import PopoverHint from '../../../../../../common/PopoverHint';
import links from '../../../../../../../common/installLinks.mjs';

class CostSavingsSection extends React.Component {
  state = {};

  onMaxHourlyPriceChange = (currentPrice, delta) =>
    Number((parseFloat(currentPrice) + delta).toFixed(2));

  render() {
    const { change, useSpotInstances, spotInstancePricing, spotInstanceMaxHourlyPrice } =
      this.props;
    return (
      <>
        <GridItem id="costsavings" md={12}>
          <FormGroup fieldId="costsavings" label="Cost saving" />
          <Field
            component={ReduxCheckbox}
            name="spot_instances"
            label="Use Amazon EC2 Spot Instances"
            description="You can save on costs by creating a machine pool running on AWS that deploys machines as
            non-guaranteed Spot Instances.  This cannot be changed after machine pool is created."
          />
        </GridItem>

        {useSpotInstances && (
          <>
            <GridItem>
              <Field
                component={RadioButtons}
                className="ocm-c-cost-savings-price-radios"
                name="spot_instance_pricing"
                options={[
                  {
                    value: 'onDemand',
                    label: 'Use On-Demand instance price',
                    description: (
                      <>The maximum price defaults to charge up to the On-Demand Instance price.</>
                    ),
                  },
                  {
                    value: 'maximum',
                    label: (
                      <>
                        Set maximum price{' '}
                        <PopoverHint
                          headerContent="Maximum hourly price for a Spot Instance"
                          hint="This value should be lower or equal to the On-Demand Instance price. This cannot be changed after the machine pool is created."
                        />
                      </>
                    ),
                    description: <>Specify the maximum hourly price for a Spot Instance.</>,
                    extraField: spotInstancePricing === 'maximum' && (
                      <Field
                        component={NumberInput}
                        className="ocm-c-cost-savings-max-price__number_input"
                        name="spot_instance_max_hourly_price"
                        inputName="spot_instance_max_hourly_price"
                        id="spot_instance_max_hourly_price"
                        props={{
                          value: parseFloat(spotInstanceMaxHourlyPrice).toFixed(2),
                          id: 'spot_instance_max_hourly_price',
                          min: 0.01,
                          onMinus: () => {
                            change(
                              'spot_instance_max_hourly_price',
                              this.onMaxHourlyPriceChange(spotInstanceMaxHourlyPrice, -0.01),
                            );
                          },
                          onPlus: () => {
                            change(
                              'spot_instance_max_hourly_price',
                              this.onMaxHourlyPriceChange(spotInstanceMaxHourlyPrice, 0.01),
                            );
                          },
                          onChange: (e) => {
                            const newHourlyMaxPriceNum = parseFloat(e.target.value);
                            change(
                              'spot_instance_max_hourly_price',
                              this.onMaxHourlyPriceChange(newHourlyMaxPriceNum, 0),
                            );
                          },
                          widthChars: '8',
                          unit: '$ Hourly',
                          isDisabled: false,
                        }}
                      />
                    ),
                  },
                ]}
                defaultValue="onDemand"
              />
            </GridItem>
            <GridItem lg={6} />
            <GridItem>
              <Alert
                className="bottom-alert"
                variant="warning"
                title="Your Spot Instance may be interrupted at any time. Use Spot Instances for workloads that can tolerate interruptions."
                isInline
              >
                <ExternalLink href={links.AWS_SPOT_INSTANCES}>
                  Learn more about Spot instances
                </ExternalLink>
              </Alert>
            </GridItem>
          </>
        )}
      </>
    );
  }
}

CostSavingsSection.propTypes = {
  change: PropTypes.func.isRequired,
  useSpotInstances: PropTypes.bool,
  spotInstancePricing: PropTypes.string,
  spotInstanceMaxHourlyPrice: PropTypes.number,
};

CostSavingsSection.defaultProps = {
  useSpotInstances: false,
  spotInstancePricing: 'onDemand',
  spotInstanceMaxHourlyPrice: 0.01,
};

export default CostSavingsSection;
