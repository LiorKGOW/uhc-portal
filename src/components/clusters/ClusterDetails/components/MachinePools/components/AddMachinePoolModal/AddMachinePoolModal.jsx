import React, { Component } from 'react';
import get from 'lodash/get';
import PropTypes from 'prop-types';
import { Field } from 'redux-form';
import {
  Form, GridItem, Grid, FormGroup,
} from '@patternfly/react-core';

import Modal from '../../../../../../common/Modal/Modal';
import ErrorBox from '../../../../../../common/ErrorBox';
import ScaleSection from '../../../../../CreateOSDPage/CreateOSDForm/FormSections/ScaleSection/ScaleSection';
import ReduxVerticalFormGroup from '../../../../../../common/ReduxFormComponents/ReduxVerticalFormGroup';
import { checkMachinePoolName } from '../../../../../../../common/validators';
import CostSavingsSection from './CostSavingsSection';

class AddMachinePoolModal extends Component {
  componentDidMount() {
    const {
      getOrganizationAndQuota,
      machineTypes,
      getMachineTypes,
    } = this.props;
    if (!machineTypes.fulfilled && !machineTypes.pending) {
      getMachineTypes();
    }
    getOrganizationAndQuota();
  }

  componentDidUpdate(prevProps) {
    const {
      addMachinePoolResponse, clearAddMachinePoolResponse, closeModal, getOrganizationAndQuota,
    } = this.props;
    if (!prevProps.addMachinePoolResponse.fulfilled && addMachinePoolResponse.fulfilled) {
      closeModal();
      clearAddMachinePoolResponse();
      getOrganizationAndQuota();
    }
  }

  cancelAddMachinePool = () => {
    const { clearAddMachinePoolResponse, closeModal } = this.props;
    clearAddMachinePoolResponse();
    closeModal();
  };

  render() {
    const {
      submit,
      addMachinePoolResponse,
      cluster,
      pristine,
      invalid,
      organization,
      canAutoScale,
      autoscalingEnabled,
      change,
      selectedMachineType,
      autoScaleMinNodesValue,
      autoScaleMaxNodesValue,
      canUseSpotInstances,
      useSpotInstances,
      spotInstancePricing,
      spotInstanceMaxHourlyPrice,
    } = this.props;

    const billingModel = get(cluster, 'billing_model');

    const hasError = addMachinePoolResponse.error && (
      <ErrorBox message="Error adding machine pool" response={addMachinePoolResponse} />
    );

    const modalDescription = 'A machine pool is a group of machines that are all clones of the same configuration, that can be used on demand by an application running on a pod.';

    const isPending = addMachinePoolResponse.pending || (organization && organization.pending);

    return (
      <Modal
        variant="large"
        title="Add machine pool"
        onClose={this.cancelAddMachinePool}
        description={modalDescription}
        primaryText="Add machine pool"
        secondaryText="Cancel"
        onPrimaryClick={submit}
        onSecondaryClick={this.cancelAddMachinePool}
        isPrimaryDisabled={isPending || pristine || invalid}
        isPending={isPending}
      >
        <>
          {hasError}
          <Form className="control-form-cursor" onSubmit={(e) => { submit(); e.preventDefault(); }}>
            <Grid>
              <GridItem md={6}>
                <FormGroup label="Machine pool name" className="pf-u-mb-md" isRequired>
                  <Field
                    component={ReduxVerticalFormGroup}
                    name="name"
                    type="text"
                    validate={checkMachinePoolName}
                    disabled={isPending}
                  />
                </FormGroup>
              </GridItem>
              <GridItem md={6} />
              <ScaleSection
                pending={isPending}
                isBYOC={!!cluster?.ccs?.enabled}
                isMultiAz={cluster.multi_az}
                machineType={selectedMachineType}
                cloudProviderID={cluster.cloud_provider.id}
                product={cluster?.subscription?.plan?.type}
                showStorageAndLoadBalancers={false}
                minNodes={0}
                isMachinePool
                inModal
                canAutoScale={canAutoScale}
                autoscalingEnabled={autoscalingEnabled}
                change={change}
                autoScaleMinNodesValue={autoScaleMinNodesValue}
                autoScaleMaxNodesValue={autoScaleMaxNodesValue}
                billingModel={billingModel}
              />
              {/* Cost savings */}
              { canUseSpotInstances
              && (
                <>
                  <CostSavingsSection
                    useSpotInstances={useSpotInstances}
                    spotInstancePricing={spotInstancePricing}
                    spotInstanceMaxHourlyPrice={spotInstanceMaxHourlyPrice}
                    change={change}
                  />
                </>
              )}
            </Grid>
          </Form>
        </>
      </Modal>
    );
  }
}

AddMachinePoolModal.propTypes = {
  closeModal: PropTypes.func.isRequired,
  submit: PropTypes.func.isRequired,
  clearAddMachinePoolResponse: PropTypes.func.isRequired,
  addMachinePoolResponse: PropTypes.object,
  cluster: PropTypes.object.isRequired,
  getOrganizationAndQuota: PropTypes.func.isRequired,
  getMachineTypes: PropTypes.func.isRequired,
  machineTypes: PropTypes.object.isRequired,
  pristine: PropTypes.bool.isRequired,
  invalid: PropTypes.bool.isRequired,
  organization: PropTypes.object,
  selectedMachineType: PropTypes.string,
  canAutoScale: PropTypes.bool.isRequired,
  autoscalingEnabled: PropTypes.bool.isRequired,
  change: PropTypes.func.isRequired,
  autoScaleMinNodesValue: PropTypes.string,
  autoScaleMaxNodesValue: PropTypes.string,
  canUseSpotInstances: PropTypes.bool.isRequired,
  useSpotInstances: PropTypes.bool.isRequired,
  spotInstancePricing: PropTypes.string,
  spotInstanceMaxHourlyPrice: PropTypes.number,
};

AddMachinePoolModal.defaultProps = {
  addMachinePoolResponse: {},
  autoScaleMinNodesValue: '0',
  autoScaleMaxNodesValue: '0',
  spotInstancePricing: 'onDemand',
  spotInstanceMaxHourlyPrice: 0.00,
};

export default AddMachinePoolModal;
