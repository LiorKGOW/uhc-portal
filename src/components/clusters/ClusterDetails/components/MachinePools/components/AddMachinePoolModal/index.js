import { connect } from 'react-redux';
import { reduxForm, formValueSelector } from 'redux-form';
import isEmpty from 'lodash/isEmpty';

import AddMachinePoolModal from './AddMachinePoolModal';
import { canAutoScaleSelector, canUseSpotInstances } from '../../MachinePoolsSelectors';
import { closeModal } from '../../../../../../common/Modal/ModalActions';
import { getMachineTypes } from '../../../../../../../redux/actions/machineTypesActions';
import { getOrganizationAndQuota } from '../../../../../../../redux/actions/userActions';
import { addMachinePoolOrNodePool, clearAddMachinePoolResponse } from '../../MachinePoolsActions';
import { isMultiAZ } from '../../../../clusterDetailsHelper';

import {
  parseReduxFormKeyValueList,
  parseReduxFormTaints,
} from '../../../../../../../common/helpers';

const reduxFormConfig = {
  form: 'AddMachinePool',
  touchOnChange: true,
};
const reduxFormAddMachinePool = reduxForm(reduxFormConfig)(AddMachinePoolModal);

const mapStateToProps = (state, ownProps) => {
  const valueSelector = formValueSelector('AddMachinePool');

  return {
    addMachinePoolResponse: state.machinePools.addMachinePoolResponse,
    machineTypes: state.machineTypes,
    organization: state.userProfile.organization,
    selectedMachineType: valueSelector(state, 'machine_type'),
    autoscalingEnabled: !!valueSelector(state, 'autoscalingEnabled'),
    autoScaleMinNodesValue: valueSelector(state, 'min_replicas'),
    autoScaleMaxNodesValue: valueSelector(state, 'max_replicas'),
    canAutoScale: canAutoScaleSelector(state, ownProps.cluster.subscription.plan.type),
    canUseSpotInstances: canUseSpotInstances(state, ownProps.cluster.subscription.plan.type),
    useSpotInstances: valueSelector(state, 'spot_instances'),
    spotInstancePricing: valueSelector(state, 'spot_instance_pricing'),
    spotInstanceMaxHourlyPrice: valueSelector(state, 'spot_instance_max_hourly_price') || 0.01,
    initialValues: {
      name: '',
      subnet: '',
      nodes_compute: '0',
      node_labels: [{}],
      taints: [{ effect: 'NoSchedule' }],
    },
    vpcListPending: state.ccsInquiries?.vpcs?.pending,
    vpcListBySubnet: state.ccsInquiries?.vpcs?.data?.bySubnetID,
  };
};

// MachinePoolAutoscaling uses min_replicas and max_replicas, NodePoolAutoscaling uses min_replica and max_replica
const replicasFieldName = ({ isHypershiftCluster }, minOrMax) =>
  `${minOrMax}_replica${!isHypershiftCluster ? 's' : ''}`;

const mapDispatchToProps = (dispatch, ownProps) => ({
  onSubmit: (formData) => {
    const { isHypershiftCluster } = ownProps;
    const machinePoolRequest = {
      id: formData.name,
    };

    if (formData.autoscalingEnabled) {
      const minNodes = parseInt(formData.min_replicas, 10);
      const maxNodes = parseInt(formData.max_replicas, 10);
      // const isMultiAz = !isHypershiftCluster && ownProps.cluster.multi_az;

      const isMultiAvailZone = isMultiAZ(ownProps.cluster);

      machinePoolRequest.autoscaling = {
        [replicasFieldName(ownProps, 'min')]: isMultiAvailZone ? minNodes * 3 : minNodes,
        [replicasFieldName(ownProps, 'max')]: isMultiAvailZone ? maxNodes * 3 : maxNodes,
      };
    } else {
      machinePoolRequest.replicas = parseInt(formData.nodes_compute, 10);
    }

    if (isHypershiftCluster) {
      machinePoolRequest.aws_node_pool = {
        instance_type: formData.machine_type,
      };
      machinePoolRequest.subnet = formData.subnet.id;
    } else {
      machinePoolRequest.instance_type = formData.machine_type;
      if (formData.spot_instances) {
        machinePoolRequest.aws = {
          spot_market_options:
            formData.spot_instance_pricing === 'maximum' &&
            formData.spot_instance_max_hourly_price !== undefined
              ? { max_price: formData.spot_instance_max_hourly_price }
              : {},
        };
      }
    }
    const parsedLabels = parseReduxFormKeyValueList(formData.node_labels);
    const parsedTaints = parseReduxFormTaints(formData.taints);

    if (!isEmpty(parsedLabels)) {
      machinePoolRequest.labels = parsedLabels;
    }

    if (parsedTaints.length > 0) {
      machinePoolRequest.taints = parsedTaints;
    }

    dispatch(
      addMachinePoolOrNodePool(
        ownProps.cluster.id,
        machinePoolRequest,
        ownProps.isHypershiftCluster,
      ),
    );
  },
  clearAddMachinePoolResponse: () => dispatch(clearAddMachinePoolResponse()),
  closeModal: () => dispatch(closeModal()),
  getOrganizationAndQuota: () => dispatch(getOrganizationAndQuota()),
  getMachineTypes: () => dispatch(getMachineTypes()),
});

export default connect(mapStateToProps, mapDispatchToProps)(reduxFormAddMachinePool);
