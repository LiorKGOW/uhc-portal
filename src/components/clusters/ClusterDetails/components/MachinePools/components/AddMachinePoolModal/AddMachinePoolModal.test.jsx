import React from 'react';
import { shallow } from 'enzyme';

import AddMachinePoolModal from './AddMachinePoolModal';

describe('<AddMachinePoolModal />', () => {
  const closeModal = jest.fn();
  const clearAddMachinePoolResponse = jest.fn();
  const submit = jest.fn();
  const cluster = {
    managed: true,
    cloud_provider: {
      id: 'aws',
    },
    subscription: {
      plan: {
        id: 'OSD',
      },
    },
    multi_az: false,
  };
  const getOrganizationAndQuota = jest.fn();
  const getMachineTypes = jest.fn();
  const change = jest.fn();
  const pendingRequest = {
    pending: true,
    fulfilled: false,
    error: false,
  };
  const wrapper = shallow(
    <AddMachinePoolModal
      isOpen
      closeModal={closeModal}
      submit={submit}
      clearAddMachinePoolResponse={clearAddMachinePoolResponse}
      cluster={cluster}
      getOrganizationAndQuota={getOrganizationAndQuota}
      getMachineTypes={getMachineTypes}
      machineTypes={pendingRequest}
      pristine={false}
      canAutoScale
      autoscalingEnabled
      change={change}
    />,
  );
  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
