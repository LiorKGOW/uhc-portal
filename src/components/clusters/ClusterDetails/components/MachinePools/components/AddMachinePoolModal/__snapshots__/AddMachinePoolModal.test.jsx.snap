// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddMachinePoolModal /> renders correctly 1`] = `
<Modal
  description="A machine pool is a group of machines that are all clones of the same configuration, that can be used on demand by an application running on a pod."
  isPrimaryDisabled={false}
  onClose={[Function]}
  onPrimaryClick={[MockFunction]}
  onSecondaryClick={[Function]}
  primaryText="Add machine pool"
  secondaryText="Cancel"
  title="Add machine pool"
  variant="large"
>
  <Form
    className="control-form-cursor"
    onSubmit={[Function]}
  >
    <Grid>
      <GridItem
        span={6}
      >
        <FormGroup
          className="pf-u-mb-md"
          isRequired={true}
          label="Machine pool name"
        >
          <Field
            component={[Function]}
            name="name"
            type="text"
            validate={[Function]}
          />
        </FormGroup>
      </GridItem>
      <GridItem
        span={6}
      />
      <ScaleSection
        autoScaleMaxNodesValue="0"
        autoScaleMinNodesValue="0"
        autoscalingEnabled={true}
        billingModel="standard"
        canAutoScale={true}
        change={[MockFunction]}
        cloudProviderID="aws"
        handleMachineTypesChange={[Function]}
        inModal={true}
        isBYOC={false}
        isMachinePool={true}
        isMultiAz={false}
        machineType=""
        minNodes={0}
        product="OSD"
        showStorageAndLoadBalancers={false}
      />
    </Grid>
  </Form>
</Modal>
`;
